### PR TITLE
Add IoT Plug and Play GA support

### PR DESCRIPTION
**Pull request by daisukeiot**
## Purpose
- Create a new IoT Plug and Play device model (dtmi:com:Microchip:PIC_IoT_WM;1")
- Updated code to 

- Report LED states 
- Update LED code.  On/Off/Blink.  Blue = WiFi, Green = DPS/IoT Hub, Yellow = User LED, Red = Error
- Control Yellow LED through writable property/desired property
- Send telemetry with temperature sensor and light sensor
- Send events when buttons are pressed, and cumulative press count.
- Control telemetry interval through the writable property
- Read DPS ID Scope from ATCA
- Control LED and save ID Scope with ATCA with CLI
- Add reboot command
- Report Max Temperature since last reboot
- Get Min, Max, Average temperature report using command
- Re-factor debug print code to be more readable.

## Does this introduce a breaking change?

```
[ ] Yes
[ x ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ x ] Feature
[ x ] Code style update (formatting, local variables)
[ x ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ x ] Other... Please describe:
IoT Plug and Play support.

## How to Test
Run IoT Plug and Play certification test.

* Test the code
For LEDs, you may use CLI to control LEDs

## What to Check
Verify that the following are valid
Passes IoT Plug and Play certification

## Other Information
Please contact Randy Wu from Microchip

### CODE REVIEW

File: AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c [hunk 1]
```diff
@@ -161,7 +163,22 @@ void MQTT_CLIENT_iotprovisioning_connect(char* deviceID)
...
    
     const az_span global_device_endpoint = AZ_SPAN_LITERAL_FROM_STR(CFG_MQTT_PROVISIONING_HOST);
-    const az_span id_scope = AZ_SPAN_LITERAL_FROM_STR(PROVISIONING_ID_SCOPE);
+
+	// Read the ID Scope from the secure element (e.g. ATECC608A)
+	atcab_read_bytes_zone(ATCA_ZONE_DATA, ATCA_SLOT_IDSCOPE, 0, atca_id_scope, sizeof(atca_id_scope));
+	debug_printInfo("DPS: ID Scope = %s", atca_id_scope);
+
+	const az_span scopeID_parm = az_span_create_from_str((char*)atca_id_scope);
+	az_span id_scope = AZ_SPAN_FROM_BUFFER(scope_id_buf);
```
Comment (ericwolz) [inline AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c:172]: I don't understand what you are doing here. It looks like you are copying the same buffer around.

Comment (ericwolz) [reply AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c:172]: why can't line 171 be 

id_scope = az_span_create_from_str((char*)atca_id_scope);

File: AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c [hunk 2]
```diff
@@ -161,7 +163,22 @@ void MQTT_CLIENT_iotprovisioning_connect(char* deviceID)
...
+	az_span_copy(id_scope, scopeID_parm);
+	id_scope = az_span_slice(id_scope, 0, az_span_size(scopeID_parm));
+
+	if (strlen((char*)atca_id_scope) < 11)
+	{
+		debug_printError("DPS: invalid ID Scope");
```
Comment (ericwolz) [inline AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c:178]: spacing

File: AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c [hunk 3]
```diff
@@ -161,7 +163,22 @@ void MQTT_CLIENT_iotprovisioning_connect(char* deviceID)
...
+	// Read the ID Scope from the secure element (e.g. ATECC608A)
+	atcab_read_bytes_zone(ATCA_ZONE_DATA, ATCA_SLOT_IDSCOPE, 0, atca_id_scope, sizeof(atca_id_scope));
+	debug_printInfo("DPS: ID Scope = %s", atca_id_scope);
+
+	const az_span scopeID_parm = az_span_create_from_str((char*)atca_id_scope);
+	az_span id_scope = AZ_SPAN_FROM_BUFFER(scope_id_buf);
+	az_span_copy(id_scope, scopeID_parm);
+	id_scope = az_span_slice(id_scope, 0, az_span_size(scopeID_parm));
+
+	if (strlen((char*)atca_id_scope) < 11)
```
Comment (ericwolz) [inline AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c:176]: move to line 170

File: AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c [hunk 4]
```diff
@@ -33,7 +33,9 @@ pf_MQTT_CLIENT pf_mqtt_iotprovisioning_client = {
 };
 
 extern uint8_t device_id_buf[100];
+extern uint8_t scope_id_buf[100];
```
Comment (ericwolz) [inline AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c]:
```suggestion
extern uint8_t dps_id_scope_buf[100];
```

File: AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c [hunk 5]
```diff
@@ -33,7 +33,9 @@ pf_MQTT_CLIENT pf_mqtt_iotprovisioning_client = {
...
 extern az_span device_id;
+uint8_t atca_id_scope[12];
```
Comment (ericwol-msft) [inline AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c:38]:
```suggestion
uint8_t atca_scope_id[12];
```

File: AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.h [hunk 6]
```diff
@@ -9,6 +9,8 @@
 #include <stdint.h>
 #include "../platform/config/IoT_Sensor_Node_config.h"
 
+#define ATCA_SLOT_IDSCOPE 8 // Slot # in ATECC608A SE which stores the ID Scope
```
Comment (ericwolz) [inline AzureIotDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.h:12]:
```suggestion
#define ATCA_SLOT_DPS_IDSCOPE 8 // Slot # in ATECC608A SE which stores the ID Scope
```

File: AzureIotDps.X/platform/cli/cli.c [hunk 7]
```diff
@@ -58,6 +59,7 @@
                         "cli_version" NEWLINE\
                         "wifi <ssid>[,<pass>,[authType]]" NEWLINE\
                         "debug" NEWLINE\
+                        "idscope <idscope>" NEWLINE\
```
Comment (ericwol-msft) [inline AzureIotDps.X/platform/cli/cli.c:62]:
```suggestion
                        "DPS scope id <scope id>" NEWLINE\
```

File: AzureIotDps.X/platform/cli/cli.c [hunk 8]
```diff
@@ -101,7 +104,8 @@ const struct cmd commands[] =
     { "device",      get_device_id },
     { "cli_version", get_cli_version },
     { "version",     get_firmware_version },
-    { "debug",       set_debug_level }
+    { "debug",       set_debug_level },
+    { "idscope",     get_set_idscope }
```
Comment (ericwolz) [inline AzureIotDps.X/platform/cli/cli.c]:
```suggestion
    { "idscope",     get_set_dps_idscope }
```

File: AzureIotDps.X/platform/cli/cli.c [hunk 9]
```diff
@@ -314,6 +318,42 @@ static void get_firmware_version(char *pArg)
     printf("v%s\r\n\4", firmware_version_number);
 }
 
+static void get_set_idscope(char *pArg)
```
Comment (ericwolz) [inline AzureIotDps.X/platform/cli/cli.c]:
```suggestion
static void get_set_dps_idscope(char *pArg)
```

File: AzureIotDps.X/platform/cli/cli.c [hunk 10]
```diff
@@ -314,6 +318,42 @@ static void get_firmware_version(char *pArg)
...
+        dps_param[params] = pch;
+        params++;
+        pch = strtok (NULL, ",");
+    }
+
+    if (dps_param[0] == NULL)
+    {
+        // Read the ID Scope from the secure element (e.g. ATECC608A)
+        atcab_read_bytes_zone(ATCA_ZONE_DATA, ATCA_SLOT_IDSCOPE, 0, (uint8_t*)atca_id_scope, sizeof(atca_id_scope));
+        printf("ID Scope = %s\r\n", atca_id_scope);
```
Comment (ericwolz) [inline AzureIotDps.X/platform/cli/cli.c]:
```suggestion
        printf("DPS ID Scope = %s\r\n", atca_id_scope);
```

File: AzureIotDps.X/platform/cli/cli.c [hunk 11]
```diff
@@ -314,6 +318,42 @@ static void get_firmware_version(char *pArg)
...
+    }
+    else
+    {
+        atca_id_scope[11] = '\0';
+        // Can execute this once to write a default ID Scope to the secure element
+        printf("Setting ID Scope = %s\r\n", dps_param[0]);
```
Comment (ericwolz) [inline AzureIotDps.X/platform/cli/cli.c]:
```suggestion
        printf("Setting DPS ID Scope = %s\r\n", dps_param[0]);
```

File: AzureIotPnpDps.X/main.c [hunk 12]
```diff
@@ -7,86 +7,346 @@
...
+
+// getMaxMinReport
+static const az_span maxmin_command_name_span = AZ_SPAN_LITERAL_FROM_STR("getMaxMinReport");
+static const az_span maxmin_max_temp_name_span = AZ_SPAN_LITERAL_FROM_STR("maxTemp");
+static const az_span maxmin_min_temp_name_span = AZ_SPAN_LITERAL_FROM_STR("minTemp");
+static const az_span maxmin_avg_temp_name_span = AZ_SPAN_LITERAL_FROM_STR("avgTemp");
+static const az_span maxmin_start_time_name_span = AZ_SPAN_LITERAL_FROM_STR("startTime");
+static const az_span maxmin_end_time_name_span = AZ_SPAN_LITERAL_FROM_STR("endTime");
+static const az_span command_empty_payload_maxmin_span = AZ_SPAN_LITERAL_FROM_STR("Payload Empty. Please specify parameter since");
+static const az_span maxmin_start_time_not_found_span = AZ_SPAN_LITERAL_FROM_STR("Start Time not round");
```
Comment (ericwolz) [inline AzureIotPnpDps.X/main.c]:
```suggestion
static const az_span maxmin_start_time_not_found_span = AZ_SPAN_LITERAL_FROM_STR("Start Time not found");
```

File: AzureIotPnpDps.X/main.c [hunk 13]
```diff
@@ -7,86 +7,346 @@
...
+	az_json_writer* jw,
+	az_span property_name,
+	az_span property_val)
+{
+  RETURN_IF_AZ_RESULT_FAILED(az_json_writer_append_property_name(jw, property_name));
+  RETURN_IF_AZ_RESULT_FAILED(az_json_writer_append_string(jw, property_val));
+  return AZ_OK;
+}
+
+/**********************************************
```
Comment (ericwolz) [inline AzureIotPnpDps.X/main.c:213]: remove dead code

File: AzureIotPnpDps.X/main.c [hunk 14]
```diff
@@ -114,94 +374,78 @@ static int mqtt_publish_message(char* topic, az_span payload, int qos)
 
   if (MQTT_CreatePublishPacket(&cloudPublishPacket) != true)
   {
-    debug_printError("MQTT: Connection lost PUBLISH failed");
+  	debug_printError("MQTT  : Connection lost PUBLISH failed");
```
Comment (ericwol-msft) [inline AzureIotPnpDps.X/main.c:377]:
```suggestion
  	debug_printError("MQTT : Connection lost PUBLISH failed");
```

File: AzureIotPnpDps.X/main.c [hunk 15]
```diff
@@ -114,94 +374,78 @@ static int mqtt_publish_message(char* topic, az_span payload, int qos)
...
-  if (az_result_failed(
-          rc = az_iot_hub_client_methods_response_get_publish_topic(
-              &hub_client,
-              request->request_id,
-              status,
-              commands_response_topic,
-              sizeof(commands_response_topic),
-              NULL)))
+  if (az_result_failed(az_iot_hub_client_methods_response_get_publish_topic(
+			                  &hub_client,
```
Comment (ericwolz) [inline AzureIotPnpDps.X/main.c]: there seams to be spacing issues thru out this PR

File: AzureIotPnpDps.X/main.c [hunk 16]
```diff
@@ -213,349 +457,693 @@ static az_result invoke_getMaxMinReport(az_span payload, az_span response, az_sp
...
-  bool ret = false;
-  if (current_device_temp > device_max_temp)
+  if (twin_properties == NULL)
   {
-    device_max_temp = current_device_temp;
-    ret = true;
+    twin_properties_ptr = &twin_properties_local;
   }
-  if (current_device_temp < device_min_temp)
+  else {
```
Comment (ericwol-msft) [inline AzureIotPnpDps.X/main.c:854]:
```suggestion
  else 
  {
```

File: AzureIotPnpDps.X/platform/cli/cli.c [hunk 17]
```diff
@@ -101,7 +106,26 @@ const struct cmd commands[] =
     { "device",      get_device_id },
     { "cli_version", get_cli_version },
     { "version",     get_firmware_version },
-    { "debug",       set_debug_level }
+    { "debug",       set_debug_level },
+    { "led",         set_led },
+    { "idscope",     get_set_idscope }
```
Comment (ericwolz) [inline AzureIotPnpDps.X/platform/cli/cli.c]:
```suggestion
    { "idscope",     get_set_dps_idscope }
```

File: AzureIotPnpDps.X/platform/cli/cli.c [hunk 18]
```diff
@@ -314,6 +344,223 @@ static void get_firmware_version(char *pArg)
...
+                    LED_stopBlinkingAndSetRed(LED_OFF);
+                    break;
+            }
+            break;
+        default:
+            break;
+    }
+}
+
+static void get_set_idscope(char *pArg)
```
Comment (ericwolz) [inline AzureIotPnpDps.X/platform/cli/cli.c]:
```suggestion
static void get_set_dps_idscope(char *pArg)
```

File: AzureIotPnpDps.X/platform/cloud/wifi_service.c [hunk 19]
```diff
@@ -244,14 +271,14 @@ static void wifiCallback(uint8_t msgType, void* pMsg)
...
-
 	case M2M_WIFI_RESP_PROVISION_INFO:
 	{
 		tstrM2MProvisionInfo* pstrProvInfo = (tstrM2MProvisionInfo*)pMsg;
+
 		if (pstrProvInfo->u8Status == M2M_SUCCESS)
 		{
 			sprintf((char*)authType, "%d", pstrProvInfo->u8SecType);
-			debug_printInfo("%s", pstrProvInfo->au8SSID);
+			debug_printInfo("ddddd %s", pstrProvInfo->au8SSID);
```
Comment (ericwolz) [inline AzureIotPnpDps.X/platform/cloud/wifi_service.c]: ???

File: AzureIotPnpDps.X/platform/led.c [hunk 20]
```diff
@@ -26,33 +26,39 @@
...
-static bool ledYellowHeld = false;
-static bool isRedLedFlashing = false;
+#define LEDS_TEST_INTERVAL  50L
+#define LED_100ms_INTERVAL	timeout_mSecToTicks(100L)
+#define LED_200ms_INTERVAL	timeout_mSecToTicks(200L)
+#define LED_500ms_INTERVAL	timeout_mSecToTicks(500L)
 
-static uint32_t redblink_task(void* payload);
-static timerStruct_t redBlink_timer = { redblink_task };
+// static bool ledYellowHeld = false;
```
Comment (ericwolz) [inline AzureIotPnpDps.X/platform/led.c]: remove this

**Pull request closed without merging**